docs/cheaha/slurm/slurm_tutorial.md

```diff
@@ -0,0 +1,528 @@
+
+       JobID    JobName  Partition    Account  AllocCPUS      State ExitCode 
+------------ ---------- ---------- ---------- ---------- ---------- -------- 
+26127143          numpy    express      USER          1  COMPLETED      0:0 
+26127143.ba+      batch                 USER          1  COMPLETED      0:0 
+26127143.ex+     extern                 USER          1  COMPLETED      0:0
+```
+
+## Example 3: Parallel Jobs
+
+Multiple jobs or tasks can be executed simultaneously using `srun` within a single batch script. In this example, the same executable `python_script_new.py` is run in parallel with distinct inputs (line 17-19). The `&` symbol at the end of each line run these commands in background. The `wait` command (line 20) performs synchronization and ensures that all background processes and parallel tasks are completed before finishing. In Line 4, three tasks are requested as there are three executables to be run in parallel. Copy the batch script into a file named `multijob.job`. Use the `conda` environemnt `pytools-env`, as shown in [example2](../slurm/slurm_tutorial.md/#example-2-sequential-job).
+
+```py linenums="1"
+#!/bin/bash
+#SBATCH --job-name=multijob
+#SBATCH --nodes=1
+#SBATCH --ntasks=3
+#SBATCH --cpus-per-task=1
+#SBATCH --mem=1G
+#SBATCH --partition=express
+#SBATCH --time=01:00:00
+#SBATCH --output=%x_%j.out
+#SBATCH --error=%x_%j.err
+
+### Loading Anaconda3 module to activate `pytools-env` conda environment
+module load Anaconda3
+conda activate pytools-env
+
+### Runs the script `python_test.py` in parallel with distinct inputs and ensures synchronization
+srun --nodes=1 --ntasks=1 python python_script_new.py 1 100000 &
+srun --nodes=1 --ntasks=1 python python_script_new.py 100001 200000 &
+srun --nodes=1 --ntasks=1 python python_script_new.py 200001 300000 &
+wait
+```
+
+Copy the following python script and call it as `python_script_new.py`. The input file takes two command-line arguments i.e. the `start` and `end` values. The script uses these values to creates an array and compute the sum of its elements using numpy. The above batch script runs three parallel instances of this Python script with different inputs.
+
+```py linenums="1"
+import sys
+import numpy as np
+
+### Verify if correct number of command-line arguments is provided
+if len(sys.argv) != 3:
+  print("Usage: python python_script_new.py <start> <end>")
+  sys.exit(1)
+
+### Passing start and end values from command-line arguments
+start = int(sys.argv[1])
+end = int(sys.argv[2])
+
+### Create an array from start to end using numpy
+input_array = np.arange(start, end)
+
+### Perform addition on the array elemnts using numpy's sum function
+sum_result = np.sum(input_array)
+
+### Print Input Range and Sum
+print("Input Range: {} to {}, Sum: {}".format(start, end, sum_result))
+```
+
+The below ouptut shows that each line corresponds to the output of one parallel execution of python script with specific input ranges.
+
+```bash
```

Comment:
Please discuss the non-intuitive "out of order" results. Execution completion doesn't happen in the same order the jobs are submitted.

Reply:
Discussed the result being out-of-order and possible approach to display the results in order.

Reply:
Looks great, thanks!

Please change "in out-of-order" to "out of order".

I would also change "If you want the results to be in order, ..." to something like "If the results must be in order to be correct, you will need to modify...". As you know, forcing an order gets rid of the parallelism. Either they're in order and not parallel, or out of order and parallel.

Of course, it may be possible to sort the results, if someone wants both :)

Reply:
Sure. Incorporated the changes.